puts/harness/src/lib.rs

```diff
@@ -0,0 +1,113 @@
+mod bindings {
+    #![allow(non_upper_case_globals)]
+    #![allow(non_camel_case_types)]
+    #![allow(non_snake_case)]
+    #![allow(improper_ctypes)]
+    #![allow(dead_code)]
+    #![allow(clippy::all)]
+    include!(env!("RUST_BINDINGS_FILE"));
+}
+
+pub use bindings::*;
+
+mod init {
+    include!(env!("RUST_PUTS_INIT_FILE"));
+}
+use std::io;
+
+pub use init::*;
+use libc::{c_char, c_void};
+use puffin::error::Error;
+
+pub type FnRegister = extern "C" fn(put: *const C_PUT_TYPE) -> ();
+
+macro_rules! define_extern_c_log {
+    ( $level:ident, $name:ident ) => {
+        unsafe extern "C" fn $name(message: *const c_char) {
+            log::log!(log::Level::$level, "{}", to_string(message));
+        }
+    };
+}
+
+define_extern_c_log!(Error, c_log_error);
+define_extern_c_log!(Warn, c_log_warn);
+define_extern_c_log!(Info, c_log_info);
+define_extern_c_log!(Debug, c_log_debug);
+define_extern_c_log!(Trace, c_log_trace);
+
+#[no_mangle]
+pub static TLSPUFFIN: C_TLSPUFFIN = C_TLSPUFFIN {
+    error: Some(c_log_error),
+    warn: Some(c_log_warn),
+    info: Some(c_log_info),
+    debug: Some(c_log_debug),
+    trace: Some(c_log_trace),
+    make_result: Some(make_result),
+};
+
+pub unsafe fn to_string(ptr: *const c_char) -> String {
+    use std::ffi::CStr;
+
+    if ptr.is_null() {
+        return "".to_owned();
```

Comment:
```suggestion
        return "NULL".to_owned();
```

Reply:
I want to keep the empty string here, because it makes the C code much simpler in many cases when passing strings to Rust. If you are uncomfortable about this, we can also return a `Result` from `to_string` and let the caller decide what to do.